src/main/flight/gps_rescue.c

```diff
@@ -567,6 +570,22 @@ static void sensorUpdate(void)
     rescueState.sensor.distanceToHomeM = rescueState.sensor.distanceToHomeCm / 100.0f;
     rescueState.sensor.groundSpeedCmS = gpsSol.groundSpeed; // cm/s
 
+    // values to modify dcm_kp to a low value when climbing, and a higher value when there is a velocity error
+    if (gpsRescueConfig()->rescueGroundspeed) {
+        const float setGroundspeedInv = 1.0f / gpsRescueConfig()->rescueGroundspeed;
+        float groundspeedErrorRatio = fabsf(rescueState.sensor.groundSpeedCmS - rescueState.sensor.velocityToHomeCmS) * setGroundspeedInv;
```

Comment:
It's true that a cosine would be better, but cos is a lot more expensive to compute, and the application here doesn't require that level of precision to effectively solve the problem.

`fobsf` is intended.  It returns the absolute error between the velocity to home and the groundspeed.  Groundspeed is always positive, but velocity to home can be negative.  It is very unusual for the velocity to home to be greater than the set groundspeed, since when the IMU is oriented accurately, the two are very close to being identical, and when the IMU is oriented incorrectly, the velocity to home is almost always less than the groundspeed, in the rescue context.  The intent is to return a value such that when the velocity to home and the groundspeed are the same, we return 0; when they differ, we return the magnitude of the difference as a positive number, expressed as a fraction of the set groundspeed.

The `groundspeedErrorRatio` value then represents absolute error between the velocity to home, and groundspeed, expressed as a fraction of the set return speed, with the lowest possible value being 1.  When there is no error, the value is 1; and it increases as the error increases, to a max of 5.  It is a simple scaled velocity to home error indicator.  

The 'normal' situation would be quad flying to home, with the IMU oriented in the pitch angle.  The velocity to home would equal the GPS groudspeed, and both would be equal to, or less than, the set groundspeed.  In this case, the `groundspeedErrorRatio` is very close to 1.0.  As a result, dcm_kp is unchanged, and the pitch angle limit is normal.  The result is the normal pitch response while the IMU is oriented correctly, and an IMU that adapts heading relatively slowly.

Let's imagine a rescue like in the video above where the IMU has oriented itself 90 degrees wrong.  The rescue has started, the speed over ground is stable at the set return velocity, but the quad is flying sideways, at 90 degrees relative to home, with zero velocity to home.  The error factor would then be 2.0 (1.0 + (1-0)/1).  This can only can only be a stable condition in a rescue when the IMU is wrong.  

Without this PR, a quad in this situation would fly sideways and take maybe 6 seconds to correct the IMU.  The PIDs target velocity to home, the quad would probably assume a high angle, and gain a high velocity, still at 90 degrees to home, taking it even further from home.  By the time the IMU has corrected itself, the quad is likely to be a long way from home, but with a bit of luck it will eventually get home.

With this PR, we use the  the `groundspeedErrorRatio` to boost dcm_kp by a factor of 2, which halves the time it takes the IMU to correct itself.  We also use the inverse of the `groundspeedErrorRatio` to cut the max allowed pitch angle by 50%, reducing forward velocity.  Should that be insufficient, and should the quad speed up even more, the dcm_kp value will increase more, and the maximum pitch angle limit will get smaller.  The result is a kind of speed limiter.  We get some speedup, but not an excessive amount.  The IMU corrects much more quickly, and without speeding off so fast, the quad rotates to home with a much tighter radius.  This can be seen in the videos above.

As the quad rotates to point more truly to home, the `groundspeedErrorRatio` falls.  This allows the quad to make a higher pitch angle if it needs to, and the dmc_kp value eases back towards normal.  After a short time we are back in the normal situation, as above, with groundspeed and velocity to home being comparable, and the velocity accurately controlled by the PIDs.

An even more difficult situation is where there is an IMU error of 180 degrees and the quad is flying 180 degrees away from home.

Imagine it has prior drift at 3x the set groundspeed, away from home, when the rescue starts.  The `groundspeedErrorRatio` would be 4.0 in this case.  This indicates a very large error between the IMU orientation and the true orientation of the quad.  We use this high error value to boost dcm_kp by four times, reducing its time constant, or lag, by a factor of 4, leading to much more rapid correction than would otherwise occur, reducing the amount of time spent flying the wrong way, and tightening the radius of the correcting turn.  At the same time, the max pitch angle is constrained to only 25% of normal, which reduces the speed away from home considerably, tightening the radius of the turn and limiting the distance traveled away.  Instead of flying hundreds of meters away and potentially failing the sanity check, the quad turns relatively quickly and gets back on track.  This is seen in the videos above.

That's the intent of the code.  There is an arbitrary element to it, I accept.  At the same time, it makes sense, and is relatively simple.

I have grappled with this for quite a long time, and was unable to find a better solution.  I explored solutions based on IMU vs GPS heading differences, simple velocity limiters, and other methods, and none were as effective, and most caused problems in edge case scenarios.

At some point, when we have position control and the ability to use position control to fly between known waypoints, all this code should be replaced.  But for now, this is the best solution I can think of, and it really does solve a problem that has been around for a very, very long time, in a reasonably simple manner.